### PR TITLE
Remove Prometheus duration prefix, remove duplicate method

### PR DESCRIPTION
**Pull request by Sean-Holcomb**
- Refactored time.go so that I could delete duplicates of methods that it contained and moved from package util to timeutil
- Added tests for ParseDuration
- Create method to remove Prometheus "offset " prefix from stings and added to some methods
- Replace string parameters with duration parameter

CS PR: https://github.com/kubecost/kubecost-cost-model/pull/458

Testing: 
Tested on GCP cluster, removing error messages

```
Error returned to client: error parsing offset (offset 3h): error parsing duration: offset 3h did not match expected format [0-9+](s|m|d|h)
```

and

```
Error] Asset ETL: ComputeAssets: clusterManagementQuery: Prometheus communication error: 400 (Bad Request) URL: 'http://kubecost-thanos-query-frontend-http.kubecost:10902/api/v1/query?max_source_resolution=5m&query=sum_over_time%28%28avg%28kubecost_cluster_management_cost%29+by+%28cluster_id%29%29%5B0m%3A1m%5D+offset+180m%29+%2A+0.016667' Headers: '{ Access-Control-Allow-Origin: [ * ], Content-Length: [ 103 ], Content-Type: [ application/json ], Date: [ Fri, 25 Jun 2021 17:00:29 GMT ], Vary: [ Accept-Encoding ], Access-Control-Allow-Headers: [ Accept, Accept-Encoding, Authorization, Content-Type, Origin ], Access-Control-Allow-Methods: [ GET, OPTIONS ], Access-Control-Expose-Headers: [ Date ], Cache-Control: [ no-store ] }', Body: '{"status":"error","errorType":"bad_data","error":"1:71: parse error: duration must be greater than 0"}
```
 and confirmed that affected endpoint clusterCostsOverTime and clusterCoster handle the offset and window parameters in the same way as they did before

### CODE REVIEW

File: pkg/util/mapper/mapper.go
Comment (michaelmdresser) [inline pkg/util/mapper/mapper.go:534]: Should leading whitespace be trimmed before trimming the prefix? I have no idea, not sure how this is used.

File: pkg/util/time.go
Comment (michaelmdresser) [inline pkg/util/time.go:71]: Likewise on trimming whitespace before.

**Review by michaelmdresser — APPROVED** [review]
A couple of small questions. Did you test to make sure cluster-sizing is working?

**Comment by nikovacevic** [comment]
Echoing @michaelmdresser, it'd be good to have an idea of how this was tested. (IMO this is a great candidate for unit testing! Shame on me for not writing them in the first place for this function.) Otherwise, it all looks good to me.

One question: did you consider adding these checks to `ParseTimeRange(duration, offset string)` instead? If so, what are the tradeoffs you considered? The reason I ask is that in that function, we know that one of the two represents an offset, so we can be forgiving if someone passes "offset 2d" instead of "2d". In `ParseDuration(duration string)`, it's a bit more strange IMO to check for the word "offset", as it's a much more general function. Curious to hear your thoughts, as the author of the fix!

**Comment by AjayTripathy** [comment]
Let's make sure this is tested, merge this to master, and patch please! Thanks all.

File: pkg/costmodel/cluster.go
Comment (Sean-Holcomb) [inline pkg/costmodel/cluster.go:75]: Cleaning offset here because it is used later in this method

File: pkg/costmodel/cluster.go
Comment (Sean-Holcomb) [inline pkg/costmodel/cluster.go:707]: Cleaning offset here because it is used later in this method

File: pkg/util/timeutil/timeutil.go
Comment (Sean-Holcomb) [inline pkg/util/timeutil/timeutil.go:190]: This is the main new method

File: pkg/costmodel/cluster.go
Comment (mbolt35) [inline pkg/costmodel/cluster.go:704]: Looping in @nikovacevic here because I know he normalize a lot of the window/offset input, but should we aim to accept a `time.Duration` in the non-http handler functions? With the motivation to remove the awkward inclusion of a `timeutil.CleanDurationString` call? The http handlers can use `queryParams.GetDuration(...)` (which should include CleanDurationString internally), and the rest of the APIs can just reliably accept a `time.Duration`. Appreciate thoughts from @michaelmdresser and @Sean-Holcomb as well.

Comment (Sean-Holcomb) [reply pkg/costmodel/cluster.go:704]: I'm onboard with that idea. I already felt like I was getting out of scope for this bug fix with the package change, but that sounds like the right approach to something I felt was a little awkward too. Just sounds like some more hack and slash clean up.

Comment (nikovacevic) [reply pkg/costmodel/cluster.go:704]: Agree that `timeutil.CleanDurationString` is awkward, and requires every engineer both to know and to remember to call that function all the time. Ideally this kind of thing would be patched in one place and "just work." (For instance, in a previous review of this PR, I asked why this logic couldn't simply be added to the beginning of `ParseTimeRange` for only the `offset string` parameter, and never got a response, but the question stands.)

I'm 100% in favor of @mbolt35's recommendation of only allowing `window` and `offset` to be passed as strings where required by the circumstance (e.g. HTTP parameters) and to be of type `time.Duration` everywhere else. That said, I also don't want to blow up scope.

@Sean-Holcomb in your judgment, is this a pervasive and serious enough issue to warrant a deep clean up effort right now? Or can we fix the originally-scoped bug with a cleaner, simpler implementation and leave deep-cleaning refactors of duration and offset strings for a future clean-up task?

Comment (mbolt35) [reply pkg/costmodel/cluster.go:704]: I think it's worth the time if you're already in there - This already makes a big difference IMO - the duplication of this method in a few places was starting to become unwieldy.

Comment (Sean-Holcomb) [reply pkg/costmodel/cluster.go:704]: @nikovacevic sorry for not responding to that comment, but you should know that it did lead me to delete some more duplicate code. Moving it ParseTimeRange method doesn't really change the need for the awkward placements at higher up in the call chain, because the string itself is causing problems when it is used for Prometheus queries it becomes "offset offset 3h". Both of this methods return a time that is used in a different way

**Comment by mbolt35** [comment]
Thing change is long overdue, thank you @Sean-Holcomb for taking it on. Great work! 

Generally, do we feel like we've covered the main use-cases in our tests? I agree with @nikovacevic that this is a perfect example of functionality we should have have good unit test coverage for.

File: pkg/costmodel/cluster.go
Comment (michaelmdresser) [reply pkg/costmodel/cluster.go]: Also strongly in favor of parsing strings into useful types in advance so we don't have to later. I'll defer to @Sean-Holcomb to determine the scoping of immediately changing this across the codebase vs. delaying.

File: pkg/util/timeutil/timeutil.go
Comment (michaelmdresser) [inline pkg/util/timeutil/timeutil.go:1]: @nikovacevic @mbolt35 I don't really want to block on this, but this is once again an instance of package naming being really annoying. I think it would be awesome to name this package `time` and then use `util.time.Foo` but darn Go would make us `import timeutil "github.com/kubecost/cost-model/pkg/util/time"` instead pretty much everywhere. What do you think of, in general, this kind of package naming to avoid overloading stdlib stuff vs. aliasing imports?

Comment (mbolt35) [reply pkg/util/timeutil/timeutil.go:1]: I am fine either way -- in general, I suggested the `timeutil` package because go conventions use `http` as an example. They say "don't name your package `util` and don't name it `http` (stdlib collision), but name it `httputil`" -- I, of course, agree with you, but I am just defaulting to their suggestions in the event that it somehow makes life easier. I don't know though :(

Comment (mbolt35) [reply pkg/util/timeutil/timeutil.go:1]: https://blog.golang.org/package-names

"Avoid meaningless package names. Packages named util, common, or misc provide clients with no sense of what the package contains. This makes it harder for clients to use the package and makes it harder for maintainers to keep the package focused. Over time, they accumulate dependencies that can make compilation significantly and unnecessarily slower, especially in large programs. And since such package names are generic, they are more likely to collide with other packages imported by client code, forcing clients to invent names to distinguish them."

and ...

"Avoid unnecessary package name collisions. While packages in different directories may have the same name, packages that are frequently used together should have distinct names. This reduces confusion and the need for local renaming in client code. For the same reason, avoid using the same name as popular standard packages like io or http."

Comment (nikovacevic) [reply pkg/util/timeutil/timeutil.go:1]: I believe @mbolt35 found a Go-related resource that espoused the idea of naming packages, e.g. `httputil` instead of `util`. Which is an idea I like, for the sake of specificity. So I actually think the right move is to name the package `timeutil`, and to refactor the rest of `pkg/util` into `pkg/util/xyzutil`. What do you think about that idea?

File: pkg/cloud/gcpprovider.go
Comment (nikovacevic) [inline pkg/cloud/gcpprovider.go:138]: Why does this use `timeutil.DurationToPromString`, but for `window` below we use `timeutil.DurationString`?

File: pkg/costmodel/aggregation.go
Comment (nikovacevic) [inline pkg/costmodel/aggregation.go:1754]: This looks suspicious.

Comment (nikovacevic) [reply pkg/costmodel/aggregation.go:1754]: Was this particular path tested? Seems difficult, perhaps, but also we should be extremely careful with sweeping changes like the ones in this PR because if little errors like this one.

File: pkg/costmodel/aggregation.go
Comment (nikovacevic) [inline pkg/costmodel/aggregation.go:1751]: At a glance, I'm not sure I understand what the function `DayDurationToHourDuration` would do. Convert from a `string` to a `time.Duration`? `string` to `string`?

File: pkg/util/timeutil/timeutil.go
Comment (nikovacevic) [inline pkg/util/timeutil/timeutil.go:70]: If I give this function `duration = 30 * time.Minute`, then I get back the string `"offset 30m"`? FWIW the name `DurationToPromString` is misleading here, as not all Prometheus durations need to be prepended with "offset", which also makes it confusing to read the code where this function is used.

Comment (nikovacevic) [reply pkg/util/timeutil/timeutil.go:70]: While I'm open to being convinced otherwise, I tend to think we don't need an entire separate function for prepending "offset" on a duration string.

File: pkg/util/timeutil/timeutil.go
Comment (nikovacevic) [inline pkg/util/timeutil/timeutil.go:55]: Good catch!

File: pkg/costmodel/cluster.go
Comment (nikovacevic) [inline pkg/costmodel/cluster.go:773]: This looks suspicious, unless `timeutil.DurationToPromString` happens to add the string "offset " to the given duration, which would be unexpected.

**Review by nikovacevic — COMMENTED** [review]
Several comments to address. In particular, there is at least one spot where I believe a bug may be introduced, so we'll want to double-check that.

In general, I'm torn about these kinds of changes. On the one hand, I agree in principle with everyone commenting here that it'd have been better to design all these functions to take `time.Duration` from the beginning. No argument here. And if we could guarantee no bugs, then I'd be excited to welcome these changes wholeheartedly. Unfortunately, I am worried about unintended consequences, especially given that many of these are old functions that are difficult to test. My conservative preference would be to leave most of them alone, only fixing the specific places we know that something is broken, or to thoroughly test every function, which I imagine would take quite some time. It might even be worth discussing whether or not to deprecate as many of these functions as possible, rather than trying to retrofit them. (e.g. do we use aggAPI cache warming any more?)

I'd love opinions from @mbolt35 and @AjayTripathy on this. Perhaps it's just that I'm sensitive because of all the bugs we've encountered this week, but this reading this PR gives me the feeling like it has the potential to do more harm than it prevents.

File: pkg/util/timeutil/timeutil.go
Comment (michaelmdresser) [reply pkg/util/timeutil/timeutil.go:1]: SGTM! I'm glad there's at least some canonical Go advice here. I wouldn't run out and refactor all of `pkg/util` but good to keep in mind for future changes like this.

File: pkg/util/timeutil/timeutil.go
Comment (Sean-Holcomb) [reply pkg/util/timeutil/timeutil.go:70]: I changed the name to DurationToPromOffsetString which I hope is a bit more clear how it is different from the other method. The function does a little more than just prepend, it also checks for empty strings. I agree that the fact that it through you off is a problem, and if you think even with the new name it isn't clear enough what it does, I will remove it. I do still like it better as a utility function.

File: pkg/costmodel/aggregation.go
Comment (Sean-Holcomb) [reply pkg/costmodel/aggregation.go:1751]: I changed the name to FormatDurationStringDaysToHours

File: pkg/costmodel/aggregation.go
Comment (Sean-Holcomb) [reply pkg/costmodel/aggregation.go]: It was not, thank you for catching, I fixes and tested the aggregate, but I am not sure that is exactly what needs to be tested with this method

**Comment by Sean-Holcomb** [comment]
@nikovacevic There is definitely the possibility that the refactoring here may cause some bugs, but the thing I think you are overlooking is that previous state of these functions did cause a bug and that is what is being fixed here. If these methods had been dealing in durations to begin with the bug never would have happened. The question is are we okay with leaving bug prone code in the codebase with some messy bandaids on top of it to avoid bugs in the short term, or is it better to risk the short term bugs to try and prevent similar situations down the line? My point here is that there is no guarantee of no bugs. 

I am also all for depreciation since it seems like we have better solutions to these problems now

**Comment by nikovacevic** [comment]
@Sean-Holcomb that makes sense, I definitely appreciate that argument. I do still think that sweeping changes tend to invite new bugs, whereas targeted changes to fix specific, reported bugs tend to have a lower chance of inviting new bugs. At any rate, I support your convictions here and it looks like we support this PR as a team. I'm happy to join Michael in approving this PR now, with the note that we should do an extra-thorough job of regression testing this release. And I hope we can revisit some of these old features with the intent to deprecate soon!

**Review by nikovacevic — APPROVED** [review]

**Pull request merged**
